lib/Dialect/TTIR/Transforms/Passes.cpp

```diff
@@ -608,6 +615,7 @@ class TTIRLayoutDPSOperandsRewriter
           createToLayoutOp(rewriter, op.getLoc(), operand.get(),
                            operandConstraint, defaultMemorySpace);
 
+      // If we changed layout (memory space or element type), update the op
       if (desiredLayout) {
```

Comment:
What's confusing to me, is why `desiredLayout` is false in this case?  We just changed layout right?

Reply:
```
%0 = tensor.empty
%1 = ttir.transpose(%arg1, %0)
%2 = tensor.empty
%3 = ttir.transpose(%0, %2)
```

From what I observed this happens:

1. Rewrite is called on `%3 = ttir.transpose(%0, %2)`
2. We update %0 layout (this updates all references of %0 including `%1 = ttir.transpose(%arg1, %0)`
3. Later on rewrite is called for `%1 = ttir.transpose(%arg1, %0)`
4. We call createLayout for %0 operand, but it's already set to layout from step 2 so we return `nullopt` and we skip update of return type

Reply:
Ok wait, sorry I just realized that I think there is a bug in the test.  The above isn't legal, the test should read:

```mlir
%0 = tensor.empty
%1 = ttir.transpose(%arg1, %0)
%2 = tensor.empty
%3 = ttir.transpose(%1, %2)
```

The _result_ (not the output arg) from DPS ops must be used in order to properly express the dependency between ops.  At the TTIR graph level, we should treat all values as though they are copied, i.e. `%0` is an argument to transpose that gets copied to `%1` and then "thrown away".  During lowering we can apply additional semantics, like move semantics from C++, where we can say that the contents from `%0` are immediately moved into `%1`.  A lowering pass that properly models tensor allocation can prove this and leverage memory aliasing, but from a graph level we need to abide by DPS semantics.

Relevant section of [MLIR docs](https://mlir.llvm.org/docs/Bufferization/#destination-passing-style):

> `outs` operands are referred to as “destinations” in the following (quotes are important as this operand isn’t modified in place but copied) and comes into place in the context of bufferization as a possible “anchor” for the bufferization algorithm. This allows the user to shape the input in a form that guarantees close to optimal bufferization result when carefully choosing the SSA value used as “destination”.
>
> For every tensor result, a DPS op has a corresponding tensor operand. If there aren’t any other conflicting uses of this tensor, the bufferization can alias it with the op result and perform the operation “in-place” by reusing the buffer allocated for this “destination” input.

Reply:
I see. Ok will check tmrw when I fix the test...